sdk/core/azure-core/src/main/java/com/azure/core/http/rest/SyncRestProxy.java

```diff
@@ -0,0 +1,692 @@
+        final Class<? extends Response<?>> cls = (Class<? extends Response<?>>) TypeUtil.getRawClass(entityType);
+
+        final HttpResponse httpResponse = response.getSourceResponse();
+        final HttpRequest request = httpResponse.getRequest();
+        final int statusCode = httpResponse.getStatusCode();
+        final HttpHeaders headers = httpResponse.getHeaders();
+        final Object decodedHeaders = response.getDecodedHeaders();
+        // Inspection of the response type needs to be performed to determine which course of action should be taken to
+        // instantiate the Response<?> from the HttpResponse.
+        //
+        // If the type is either the Response or PagedResponse interface from azure-core a new instance of either
+        // ResponseBase or PagedResponseBase can be returned.
+        if (cls.equals(Response.class)) {
+            // For Response return a new instance of ResponseBase cast to the class.
+            return cls.cast(new ResponseBase<>(request, statusCode, headers, bodyAsObject, decodedHeaders));
+        } else if (cls.equals(PagedResponse.class)) {
+            // For PagedResponse return a new instance of PagedResponseBase cast to the class.
+            //
+            // PagedResponse needs an additional check that the bodyAsObject implements Page.
+            //
+            // If the bodyAsObject is null use the constructor that take items and continuation token with null.
+            // Otherwise, use the constructor that take Page.
+            if (bodyAsObject != null && !TypeUtil.isTypeOrSubTypeOf(bodyAsObject.getClass(), Page.class)) {
+                throw LOGGER.logExceptionAsError(new RuntimeException(MUST_IMPLEMENT_PAGE_ERROR));
+            } else if (bodyAsObject == null) {
+                return (cls.cast(new PagedResponseBase<>(request, statusCode, headers, null, null,
+                    decodedHeaders)));
+            } else {
+                return (cls.cast(new PagedResponseBase<>(request, statusCode, headers, (Page<?>) bodyAsObject,
+                    decodedHeaders)));
+            }
+        }
+
+        // Otherwise, rely on reflection, for now, to get the best constructor to use to create the Response sub-type.
+        //
+        // Ideally, in the future the SDKs won't need to dabble in reflection here as the Response sub-types should be
+        // given a way to register their constructor as a callback method that consumes HttpDecodedResponse and the
+        // body as an Object.
+
+        MethodHandle ctr = RESPONSE_CONSTRUCTORS_CACHE.get(cls);
+
+        if (ctr == null) {
+            throw new RuntimeException("Cannot find suitable constructor for class " + cls);
+        }
+        return RESPONSE_CONSTRUCTORS_CACHE.invokeSync(ctr, response, bodyAsObject);
+    }
+
+    private Object handleBodyReturnTypeSync(final HttpDecodedResponse response,
```

Comment:
Another nit: as a whole having this second RestProxy-type class makes me think we should tackle https://github.com/Azure/azure-sdk-for-java/issues/10398 as part of sync stack